File: c++/medium/560.subarrays_sum_equals_k/count_subarrays_sum_equals_k.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        if(nums.size() == 1) {
            return nums[0] == k;
        }

        vector<int> prefixSum;
        prefixSum.reserve(nums.size());
        prefixSum.push_back(nums[0]);

        unordered_map<int, int> sumCounter;
        sumCounter[0] = 1;
        sumCounter[nums[0]] += 1;
        
        int total = 0;
        if(nums[0] == k) total++;

        for(int i = 1; i < nums.size(); i++) {
            int current = prefixSum[i-1] + nums[i];

            if(sumCounter.find(current - k) != sumCounter.end()) {
                total += sumCounter[current - k];
            }

            prefixSum.push_back(current);
            sumCounter[current]++;
        }
        
        return total;
    }
};

File: python/medium/560.subarrays_sum_equals_k/count_subarrays_sum_equals_k.py
from typing import List

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        if len(nums) == 1:
            return int(nums[0] == k)

        total = 1 if nums[0] == k else 0
        sumCounter = {
            0: 1,
        }
        sumCounter[nums[0]] = sumCounter.get(nums[0], 0) + 1 # in case k and first element = 0
        prefixSum = []
        prefixSum.append(nums[0])

        for i in range(1, len(nums)):
            current = prefixSum[i-1] + nums[i]
            prefixSum.append(current)
            if current-k in sumCounter:
                total += sumCounter[current-k]

            sumCounter[current] = sumCounter.get(current, 0) + 1

        return total
        